### PR TITLE
Update and test server selection logic for $out/$merge aggregation pipelines

### PR DESCRIPTION
**Pull request by jyemin**
JAVA-4380

### CODE REVIEW

**Comment by jyemin** [comment]
Implemented logic from https://github.com/mongodb/specifications/pull/1095

File: driver-core/src/main/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelector.java [hunk 1]
```diff
@@ -40,22 +41,22 @@ public ReadPreferenceWithFallbackServerSelector(final ReadPreference preferredRe
 
     @Override
     public List<ServerDescription> select(final ClusterDescription clusterDescription) {
-        if (maxWireVersion(clusterDescription) >= minWireVersion) {
-            appliedReadPreference = preferredReadPreference;
-            return new ReadPreferenceServerSelector(preferredReadPreference).select(clusterDescription);
-        } else {
+        if (hasAnyPreMinWireVersionServers(clusterDescription)) {
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelector.java:44]: Can I suggest a simplification: `clusterContainsOlderServers`

File: driver-core/src/test/unit/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelectorTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import static com.mongodb.internal.operation.ServerVersionHelper.FOUR_DOT_FOUR_WIRE_VERSION;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class ReadPreferenceWithFallbackServerSelectorTest {
+
+  @Test
+  public void shouldSelectCorrectServers() {
```
Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelectorTest.java:44]: Nit: these tests would be easier to read if they were split into individual tests.

**Review by rozza — APPROVED** [review]
I have two suggestions to potentially improve future usability. 

Either way its a LGTM from me.

File: driver-core/src/main/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelector.java (hunk 1 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelector.java]: Much clearer. Done

File: driver-core/src/test/unit/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelectorTest.java (hunk 2 above)
Comment (jyemin) [reply driver-core/src/test/unit/com/mongodb/internal/selector/ReadPreferenceWithFallbackServerSelectorTest.java]: Yeah, I was being lazy.  Done.

**Pull request merged**